### DISCUSSION TITLE
The engine is not picklable and cannot be passed to a pipe

### DISCUSSION THREAD

**Discussion by simsicon** [sqlalchemy/sqlalchemy#7308] [Usage Questions] [answered]
Hi, I have encountered an error with CPython 3.9.6 on Apple Silicon, and I have tested the code works fine on CPython 3.9.6 LInux/X86 build. Since the bug happened when using sqlalchemy, I think it's worth to mention here.

I have pushed the code to https://github.com/simsicon/multiprocessing-initializer
Please take a look to the poc.py. Thank you.

**Comment by CaselIT** [comment, marked as answer]
Hi,

This is not an issue with apple silicon. It's just that the engine (the pool actually) does not support pickle since it's connect function is a local variable. On windows I get the same error `AttributeError: Can't pickle local object 'create_engine.<locals>.connect'`.

I think with linux it works because it uses pool uses fork, while on osx (any processor, also intel) and windows instead it will spawn a new process.

So the solution is don't do that :) pass the create engine args instead and re-create the engine inside the process

**Reply by zzzeek** [reply]
does our example at https://docs.sqlalchemy.org/en/14/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork also not work on windows?   that's news to me

**Reply by zzzeek** [reply]
also osx doesnt support fork?  also news to me... https://stackoverflow.com/questions/4326684/how-to-create-a-process-on-mac-os-using-fork-and-exec

**Reply by CaselIT** [reply]
osx supports fork, but since python3.8 the default multiprocessing method in osx has been spawn

**Reply by zzzeek** [reply]
so....how does that work as far as state?  it pickles up everything that's not a module in globals() and sends it over the wire ?

**Reply by zzzeek** [reply]
"the wire" you know, the pipe

**Reply by CaselIT** [reply]
> 
> 
> does our example at https://docs.sqlalchemy.org/en/14/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork also not work on windows? that's news to me

No that works, because it does not need to pickle anything, the spwaned process will re-import the modude and create a new engine instance

**Reply by CaselIT** [reply]
> 
> 
> so....how does that work as far as state? it pickles up everything that's not a module in globals() and sends it over the wire ?

I guess so, yes.

**Reply by CaselIT** [reply]
I guess for osx you can change the multiprocessing method to `fork` using https://docs.python.org/3/library/multiprocessing.html?highlight=multiprocessing#multiprocessing.set_start_method

this will of course not work on windows

**Reply by CaselIT** [reply]
> 
> 
> "the wire" you know, the pipe

if you look at the trace it's trying to send stuff on a pipe:
```
  ...
  File "/Users/zheng/.pyenv/versions/3.9.6/lib/python3.9/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/Users/zheng/.pyenv/versions/3.9.6/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Users/zheng/.pyenv/versions/3.9.6/lib/python3.9/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/Users/zheng/.pyenv/versions/3.9.6/lib/python3.9/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/Users/zheng/.pyenv/versions/3.9.6/lib/python3.9/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
```

**Reply by zzzeek** [reply]
> > does our example at https://docs.sqlalchemy.org/en/14/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork also not work on windows? that's news to me
> 
> No that works, because it does not need to pickle anything, the spwaned process will re-import the modude and create a new engine instance

that code as written will run create_engine() in each process?  wow i had no idea

**Reply by CaselIT** [reply]
yes, with spawn a new empty process is created, python is run, the module is imported, then the function is run.

basically a fancy way of shelling out a `python -c 'import module; module.the_method()`

**Reply by CaselIT** [reply]
Note that in the example in the docs if the engine is passed as arg will also fail with the same error, since it would need to pickle the engine and pass it to the method call
